rs/xnet/payload_builder/src/lib.rs

```diff
@@ -1303,52 +1304,61 @@ impl PoolRefillTask {
-                            info!(
-                                log,
-                                "Failed to query stream slice for subnet {} from node {}: {}",
-                                subnet_id,
-                                endpoint_locator.node_id,
-                                e
+            query_set.spawn_on(
+                async move {
+                    metrics.outstanding_queries.inc();
+                    let query_result = xnet_client.query(&endpoint_locator).await;
+                    metrics.outstanding_queries.dec();
+                    let since = Instant::now();
+                    let proximity = endpoint_locator.proximity.into();
+
+                    match query_result {
+                        Ok(slice) => {
+                            let res = if witness_begin != msg_begin {
+                                // Pulled a stream suffix, append to pooled slice.
+                                pool.lock()
+                                    .unwrap()
+                                    .append(subnet_id, slice, registry_version, log)
+                            } else {
+                                // Pulled a complete stream, replace pooled slice (if any).
+                                pool.lock()
+                                    .unwrap()
+                                    .put(subnet_id, slice, registry_version, log)
+                            };
+                            let status = match res {
+                                Ok(()) => STATUS_SUCCESS,
+                                Err(e) => e.to_label_value(),
+                            };
+
+                            metrics.observe_query_slice_duration(status, proximity, since);
+                            metrics.observe_pull_attempt(status);
+                        }
+
+                        Err(e) => {
+                            metrics.observe_query_slice_duration(
+                                &e.to_label_value(),
+                                proximity,
+                                since,
                             );
+                            metrics.observe_pull_attempt(&e.to_label_value());
+                            if let XNetClientError::NoContent = e {
+                            } else if Self::pass_log_sampling() {
+                                info!(
+                                    log,
+                                    "Failed to query stream slice for subnet {} from node {}: {}",
+                                    subnet_id,
+                                    endpoint_locator.node_id,
+                                    e
+                                );
+                            }
                         }
                     }
-                }
-            });
+                },
+                &self.runtime_handle,
+            );
         }
+        self.runtime_handle.block_on(query_set.join_all());
```

Comment:
I really don't like (essentially) all streams blocking on one slow replica. I.e. even in the absence of high contention or bugs, it could easily happen that we only refill the pool every 5 seconds, once the slowest pull has timed out.

I'd be OK with saying "only one pull per destination replica"; or even "only one pull per destination subnet"; but not with this kind of long tail latency.

BTW, [metrics](https://grafana.mainnet.dfinity.network/explore?orgId=1&left=%7B%22datasource%22:%22P119963096DF10E67%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P119963096DF10E67%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22%20%20%20%20quantile%20by%28ic_subnet%29%20%28%5Cn%20%20%20%20%20%20%20%201,%5Cn%20%20%20%20%20%20%20%20xnet_builder_outstanding_queries%5Cn%20%20%20%20%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221726047998578%22,%22to%22:%221726051597131%22%7D%7D) show that we rarely exceed 50 concurrent XNet queries, with [the average in the low single digits](https://grafana.mainnet.dfinity.network/explore?orgId=1&left=%7B%22datasource%22:%22P119963096DF10E67%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%22P119963096DF10E67%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22%20%20%20%20quantile%20by%28ic_subnet%29%20%28%5Cn%20%20%20%20%20%20%20%20.5,%5Cn%20%20%20%20%20%20%20%20xnet_builder_outstanding_queries%5Cn%20%20%20%20%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221726047998578%22,%22to%22:%221726051597131%22%7D%7D).

Reply:
+1